### ISSUE TITLE
Favicon should not be declared!

### ISSUE THREAD

**Issue by DarthDestroyer** [pagekit/pagekit#536]
http://realfavicongenerator.net/faq#why_ico_not_declared

As you look at the Link and the Description of it - it should not be declared like it is!

**Comment by florianletsch** [comment]
Thanks for the research. But doesn't this only apply when using both an ICO and a PNG file?

Wikipedia lists these usage options. Browser support looks pretty great: https://en.wikipedia.org/wiki/Favicon#How_to_use

Also, without declaring it we could only support `*.ico` files. This way, we can also support `*.png` and `*.gif` files.

**Comment by DarthDestroyer** [comment]
like "realfavicongenerator" and wikipedia says, every browser choose whatever they like.

i would do it on the hard way! the solution of realfavicongenerator is very great! creating every option from user-input would be the best way!

the dirty way could be an user-manual override? you can choose a image (like it is) or you could overwrite the whole icon-thingy with a textbox?

but your thinking isnt false - i think.... i'm not an expert in that, what will happen.

i came to the problem because i had tried to implement the realfavicongenerator icons and got the problem, that there is a declaration...

in my opinion, it would be really great for mobile devices (cause of desktop-icons)

also, in the meta, there could be an option to paint the navbar ;) (color-input field?)

Greetings

**Comment by xorinzor** [comment]
Declared or not, if you add your <meta> or <link rel="icon"> tags to the header field in the admin panel it will override any previously defined metatags or <link rel="icon">'s with the same name.

**Comment by huglester** [comment]
When I started using FAVICONS as <link rel="icon"> - I thought it was a better idea, than using /favicon.ico.

But later, I realized that almost any search bot (or whatever) will scan for favicon.ico, ending in 404 error logs, etc. Generally 404 errors take longer to respond, are not cached, etc etc.. So I abandoned using <meta> favicons and started using the /favicon.ico

But I never tried using GIF animated favicons or so.

I generally take for example image.png and convert it to favicon, by calling custom script:
`makefavicon image.png` and it will create favicon.ico

My script:
```bash
#!/bin/bash
convert -resize x32 -gravity center -crop 32x32+0+0 $1 -flatten -colors 256 favicon.ico
```

Just my 2cents.

Jaroslav

**Comment by DarthDestroyer** [comment]
You missed some Parts, @huglester!

First of all, there has to be a favicon!
Second, it has to be in multi-size format (16x16,32x32,48x48)

But, you better not declare it in the html-meta! As far as i had understood the Page and the other Guys, who had researched the problem, it is all for multi-browser and mobile support...

Regards

**Comment by iamswain** [comment]
@xorinzor I'm not sure this really works. In my experience the browser grabs the first defined favicon, and since the Header Code gets executed after the Pagekit favicon has been added (or the one defined in settings) the browser will grab that one.

A solution would be that the favicon code will not be outputted when no favicon has been added (so Theme developers can add a favicon themselves in whatever way they like)

**Comment by xorinzor** [comment]
I've never had any issues doing it this way, that said I never thoroughly checked everything because I felt I could spend my time in a better way, it's just a favicon after all, as long as its properly displayed in the desktop browser I'm happy (mobile icon's use a different metatag and search engines don't use favicons anywhere at all)

**Comment by iamswain** [comment]
Either way, it doesn't really matter. It's just a bit unclean to have two favicon tags. Another solution would be an option to say "I display my theme's favicon, not the one from the system".

However, I'm kind of new to the Pagekit development etc. so I'm still trying to find my way through the file system.

**Comment by xorinzor** [comment]
I agree a better option would be to default to the standard favicon and have an upload-file option in the admin panel to upload a different favicon.

**Comment by iamswain** [comment]
That's what currently is happening, I'm not sure if what I'm trying to say even fits in this issue anymore but I'll try to rephrase.

In my theme I have currently about 30 HTML tags for Favicon usage (all generated by http://realfavicongenerator.net), which all have a use. However I don't need the ones from Pagekit itself due to the ones declared in my theme.

If there was a way to disable the Pagekit ones, that would allow me to create a settings page (or not) for the favicons in my theme. I do think the "disable default favicons" option would be the most realistic and most flexible. I can't expect the settings page from Pagekit to have 30 upload fields.

I hope this is a bit more understandable, and if need be, I can create a new issue for this.

**Comment by xorinzor** [comment]
No I don't think a "disable" option is a good idea.

The theme should be adjusted to the CMS, not the other way around. Pagekit in this case should provide an option to configure a different image for the favicon.
That way you won't have to mess around with favicon's in a theme anymore too.

**Comment by iamswain** [comment]
@xorinzor You're right. I'll just let this rest for now and try to get back to the real issue.

@DarthDestroyer The link you references shows how **not** to declare it: `<link rel="shortcut icon" type="image/x-icon" href="favicon.ico">`, however, if I upload a .png this is what get's outputted: `<link href="/storage/favicon.png" rel="shortcut icon" type="image/x-icon">`. Where is the error, as in, how should it be outputted in order to solve this issue?

**Comment by DarthDestroyer** [comment]
Ok...

As far as i know, the favicon should ever be a multi-size ico file! 
There is no need to use a png, because ico can handle transparency, too... 
And the multi-size icon is needed for crossbrowser support... 

The favicon itself should not be declared it has to be in the root dir.

But, like i said... im only referencing some infos that i had found! 
If you want to understand it, you can look into the links! ;-)

And im out, when i think about gif files... i dont know, if it is the same as the png way, but ico cant handle animations...

Whatever... I dont want to make a big Problem out of this "small" issue... I want to optimize the project for everyone! Not only me is interested in crossbrowser and mobile support! 

The glory way, the best, the epic, the ultimate way would be to implement something like the favicongenerator! 

You put a png in and it will be converted in all needed files, for Browser and mobiles... 

It is only a small thing, for desktops... but it is a big thing on mobiles

**Comment by iamswain** [comment]
A connection to the generator would be amazing! I did see they have an API (http://realfavicongenerator.net/api/), but I'm not sure it's a good thing since you'll be depending on them. However, you could always keep the current method as a fallback.

**Comment by DarthDestroyer** [comment]
I dont see a reason for a "Connection"... sure, you have a lot of options and Buttons... 

but at the source, its only a graphic to ico converter, resizer and echo of these html tags.

**Comment by xorinzor** [comment]
A connection with any kind of API would be something for an extension, not the core CMS

**Comment by iamswain** [comment]
@DarthDestroyer I guess this would be the ultimate solution. Basically 'copy' the generator and make it a part of Pagekit.

@xorinzor I agree with you there, but then we get back to the point that the default favicons have to be disabled for that to work. No clue if this at all is possible.

**Comment by xorinzor** [comment]
@iamswain they dont have to be disabled for that, you can like I said just add an option to configure the default favicon.

If the extension wants to add other favicons it can just add them to the <head> code

**Comment by florianletsch** [comment]
In my opinion Pagekit should provide a Favicon implementation that is

1. Simple.
2. Works for most current cases.

Anything that goes beyond these 2 scopes should be implemented by a custom theme. 

With regards to **1. Simplicity**: I do not think that Pagekit should convert uploaded files to different formats, create *.ico files with multiple resolutions and print 40 lines in the head section. If a project requires that, the developer should create these assets and create the required markup, i.e. using the excellent realfavicongenerator.

So, now what satisfies **2. works in most cases**? Potentially just an .ico file in the root directory. However, Pagekit should not be able to write files in that directory. And once we start storing files inside `/storage`, we might as well use png files. PNG is, after all, the format of the future (as [stated by the creator of the favicongenerator himself](https://css-tricks.com/favicon-quiz/)):

> Let's face an unsatisfying truth. Altough all desktop browsers support favicon.ico, this file is mostly deprecated. Yes, you can still rely on this file and it will work as expected. But seriously, it's an old deprecated format. It is not used anywhere else in web technology. PNG is much more consistent.

My suggestion: Upload a high resolution square PNG file (dimensions `Y by Y`), link it from the head section as follows:

```
<link rel="icon" type="image/png" href="/storage/favicon.png" sizes="YxY">
```

In addition, we also have the *app icon*, which will be rendered as:

```
<link href="/storage/appicon.png" rel="apple-touch-icon-precomposed">
```

What do you think about this? Am I missing something important?

**Comment by iamswain** [comment]
@florianletsch Sounds like a solid solution, what should it say in the `sizes=YxY` part? The sizes of the uploaded image or default sizes (16x16 32x32 etc.)?

**Comment by florianletsch** [comment]
@iamswain I would say the size of the uploaded image. Maybe we could say in the upload interface something like:

> Recommended resolution: 64x64 (or larger)

**Comment by DarthDestroyer** [comment]
I am totally with all of you guys! @florianletsch is not really wrong with his idea... 
But, i have to add a Sentence... maybe two!

@florianletsch is right about the writing in the root directory! but, it isnt important where it is! first of all, you should not miss that file! (for crossbrowser and whatsoever support!) and then, declare the location in the link tag.

The best size for resizing would be 200x200 pixels, while apple needs max 180x180 and something else needs 194x194 and android-chrome needs 192x192.

I bet, that only a little percentage of webpages has a multi-size ico, too! But that is another point!
It is not only the icon, that is missing... The creation of a multi-size ico is a little bit "funny", when you not use a generator like favicongenerator.

TLDR:
When you want a totally crossbrowser (mobile included) support, your Meta should like this:
`<link rel="apple-touch-icon" sizes="57x57" href="/apple-touch-icon-57x57.png">
<link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
<link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
<link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
<link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
<link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">
<link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="/favicon-194x194.png" sizes="194x194">
<link rel="icon" type="image/png" href="/favicon-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="/android-chrome-192x192.png" sizes="192x192">
<link rel="icon" type="image/png" href="/favicon-16x16.png" sizes="16x16">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#000000">
<link rel="shortcut icon" href="/favicon.ico">
<meta name="msapplication-TileColor" content="#bd2a00">
<meta name="msapplication-TileImage" content="/mstile-144x144.png">
<meta name="theme-color" content="#cf0000">`

And now, say to a user: "please create them all by hand".
I bet, that will fail! So, why complain about a middle to none Solution when it is only a little bit code to implement an image-resize function?

It is only my Opinion! I dont want to create a big trouble about such a "small feature", but it would be a big help for unexperienced people and an even larger addition compared to other CMSs!

I dont know, why every Browser has to brew his own soup... But we have to deal with it until it gets a standard... 

Until now, i have the above code in the header tag settings and it works well!
